File: 2503-maximum-number-of-points-from-grid-queries/2503-maximum-number-of-points-from-grid-queries.cpp
class Solution {
private:
    // Directions for movement (right, down, left, up)
    vector<pair<int, int>> directions{{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    int m, n;
    bool isValid(int row, int col, vector<vector<bool>>& visited){
        return 0 <= row && row < m && 0 <= col && col < n && !visited[row][col];
    }

public:
    vector<int> maxPoints(vector<vector<int>>& grid, vector<int>& queries) {
        this -> m = grid.size(), this -> n = grid[0].size();

        // Store queries along with their original indices to restore order later
        vector<pair<int, int>> sortedQueries;
        for (int i = 0; i < queries.size(); ++i)
            sortedQueries.push_back({queries[i], i});

        // Sort queries by value in ascending order
        sort(sortedQueries.begin(), sortedQueries.end());

        vector<int> ans(queries.size(), 0);
        // Keep track of the number of cells processed
        int count = 0;

        vector<vector<bool>> visited(m, vector<bool>(n, false));
        
        // Min-heap (priority queue) to process cells in increasing order of value
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> minHeap;

        // Start from the top-left cell
        minHeap.push({grid[0][0], 0, 0});
        visited[0][0] = true;

        // Process queries in sorted order
        for (const auto& [val, i] : sortedQueries){
            // Expand the cells that are smaller than the current query value
            while (!minHeap.empty() && val > minHeap.top()[0]){
                const auto it = minHeap.top();
                minHeap.pop();
                int currVal = it[0], currRow = it[1], currCol = it[2];
                // Increment count of valid cells
                ++count;

                // Explore all four possible directions
                for (auto& [dr, dc] : directions){
                    int newRow = currRow + dr, newCol = currCol + dc;

                    // Check if the new cell is within bounds and not visited
                    if (isValid(newRow, newCol, visited)){
                        // Mark as visited
                        visited[newRow][newCol] = true;
                        minHeap.push({grid[newRow][newCol], newRow, newCol});
                    }
                }
            }

            // Store the result for this query
            ans[i] = count;
        }

        return ans;
    }
};

File: 2503-maximum-number-of-points-from-grid-queries/2503-maximum-number-of-points-from-grid-queries.py
class Solution:
    def maxPoints(self, grid: List[List[int]], queries: List[int]) -> List[int]:
        m, n = len(grid), len(grid[0])
        ans = [0] * len(queries)

        # Directions for movement (right, down, left, up)
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        # Store queries along with their origial indices to restore order later
        sorted_queries = sorted([(val, idx) for idx, val in enumerate(queries)])

        # Min-heap (priority queue) to process cells in increasing order of value
        min_heap = [(grid[0][0], 0, 0)]
        visited = [[False] * n for _ in range(m)]

        # Keep track of the number of cells processed
        total_pts = 0
        # Start from the top-left cell
        visited[0][0] = True

        # Process queries in sorted order
        for val, idx in sorted_queries:
            # Expand the cells that are smaller than the current query value
            while min_heap and min_heap[0][0] < val:
                curr_val, curr_r, curr_c = heappop(min_heap)

                # Increment count of valid cells
                total_pts += 1

                # Explore all four possible directions
                for dr, dc in dirs:
                    new_r, new_c = curr_r + dr, curr_c + dc

                    # Check if the new cell is within bounds and not visited
                    if (new_r >= 0
                    and new_c >= 0
                    and new_r < m
                    and new_c < n
                    and not visited[new_r][new_c]
                    ):
                        heappush(min_heap, (grid[new_r][new_c], new_r, new_c))
                        # Mark as visited
                        visited[new_r][new_c] = True

            # Store the result for this query
            ans[idx] = total_pts
        
        return ans